### DISCUSSION TITLE
whole element visibility

### DISCUSSION THREAD

**Discussion by gureckis** [vueuse/vueuse#2292] [Q&A]
The `useElementVisibility` detects if *any* part of an element is visible in the viewport.  however, a closely related query would be if the element is completely visible in the view port?  is there an easy way to implement that using vueuse?

**Comment by curtgrimes** [comment]
Instead of using [`useElementVisibility()`](https://vueuse.org/core/useelementvisibility/), try using an intersection observer with [`useIntersectionObserver()`](https://vueuse.org/core/useintersectionobserver/). It accepts [a `threshold` option](https://vueuse.org/core/useintersectionobserver/#type-declarations) that lets you decide if you want to know if only a portion of the element is visible in a viewport (threshold < 1) or if you want to know when the entire element is visible in the viewport (threshold = 1).

For example, to use `useIntersectionObserver()` to know when an element is entirely visible in the viewport:

```vue
<script setup lang="ts">
import { ref } from 'vue';
import { useIntersectionObserver } from '@vueuse/core';

const watchedElement = ref();
const targetIsVisible = ref(false);

useIntersectionObserver(
  watchedElement,
  ([{ isIntersecting }]) => {
    // Run this callback whenever watchedElement
    // intersects the viewport boundary
    targetIsVisible.value = isIntersecting;
  },
  {
    // The entire element must become visible in
    // the viewport to trigger an intersection
    threshold: 1,
  }
);
</script>

<template>
  <div class="message">
    Scroll down to make element appear. The watched element
    <strong>{{ targetIsVisible ? 'IS' : 'IS NOT' }}</strong> completely visible
    in the viewport.
  </div>
  <div ref="watchedElement" class="watchedElement">This is the element</div>
</template>
```

Demo: https://stackblitz.com/edit/vitejs-vite-cyjfwh?file=src%2FApp.vue&terminal=dev

![CleanShot 2022-10-09 at 18 50 08](https://user-images.githubusercontent.com/3038600/194785120-432703a9-e1b0-4bf7-b4ed-c7c670d68c7b.gif)

As a bonus, this is likely more performant than using `useElementVisibility()` since `useElementVisibility()` [relies on using a window scroll event listener](https://github.com/vueuse/vueuse/blob/27b66cd738d626d6b4f11753d8325d1438753737/packages/core/useElementVisibility/index.ts#L53) (this can have performance issues).
